library/init/meta/interactive.lean

```diff
@@ -521,6 +517,25 @@ do x ← mk_fresh_name,
    t ← get_local x,
    induction (to_pexpr t) rec_name hs ([] : list name)
 
+private meta def cases_arg_p : parser (option name × pexpr) :=
+with_desc "(id :)? expr" $ do
+  t ← texpr,
+  match t with
+  | (local_const x _ _ _) :=
```

Comment:
We can't do it for `have`, because the type is not mandatory (you can do just `have h` and it is okay). You wouldn't be able to distinguish it from `have t` where `t` is a local constant. We could use this trick for `generalize`, though, with a bit of modification to also parse the equality right.